builder_test.go

```diff
@@ -334,6 +334,273 @@ func testBuilder(t *testing.T, when spec.G, it spec.S) {
+						})
+					})
+
+					when("overriding default process type, with a non-default process type", func() {
+						it.Before(func() {
+							builder.Group.Group = []lifecycle.GroupBuildpack{
+								{ID: "A", Version: "v1", API: latestBuildpackAPI.String()},
+								{ID: "B", Version: "v2", API: latestBuildpackAPI.String()},
+								{ID: "C", Version: "v3", API: latestBuildpackAPI.String()},
+							}
+						})
+
+						it("should warn and not set any default process", func() {
+							bpB := testmock.NewMockBuildpack(mockCtrl)
+							buildpackStore.EXPECT().Lookup("A", "v1").Return(bpB, nil)
+							bpB.EXPECT().Build(gomock.Any(), config).Return(lifecycle.BuildResult{
+								Processes: []launch.Process{
+									{
+										Type:        "some-type",
+										Command:     "bpA-command",
+										Args:        []string{"bpA-arg"},
+										Direct:      false,
+										BuildpackID: "A",
+										Default:     true,
+									},
+								},
+							}, nil)
+
+							bpA := testmock.NewMockBuildpack(mockCtrl)
+							buildpackStore.EXPECT().Lookup("B", "v2").Return(bpA, nil)
+							bpA.EXPECT().Build(gomock.Any(), config).Return(lifecycle.BuildResult{
+								Processes: []launch.Process{
+									{
+										Type:        "override-type",
+										Command:     "bpB-command",
+										Args:        []string{"bpB-arg"},
+										Direct:      true,
+										BuildpackID: "B",
+										Default:     true,
+									},
+								},
+							}, nil)
+
+							bpC := testmock.NewMockBuildpack(mockCtrl)
+							buildpackStore.EXPECT().Lookup("C", "v3").Return(bpC, nil)
+							bpC.EXPECT().Build(gomock.Any(), config).Return(lifecycle.BuildResult{
+								Processes: []launch.Process{
+									{
+										Type:        "override-type",
+										Command:     "bpC-command",
+										Args:        []string{"bpC-arg"},
+										Direct:      false,
+										BuildpackID: "C",
+									},
+								},
+							}, nil)
+
+							metadata, err := builder.Build()
+							if err != nil {
+								t.Fatalf("Unexpected error:\n%s\n", err)
+							}
+							if s := cmp.Diff(metadata.Processes, []launch.Process{
+								{
+									Type:        "override-type",
+									Command:     "bpC-command",
+									Args:        []string{"bpC-arg"},
+									Direct:      false,
+									BuildpackID: "C",
+								},
+								{
+									Type:        "some-type",
+									Command:     "bpA-command",
+									Args:        []string{"bpA-arg"},
+									Direct:      false,
+									BuildpackID: "A",
+								},
+							}); s != "" {
+								t.Fatalf("Unexpected:\n%s\n", s)
+							}
+
+							expected := "Warning: redefining the following default process type with a process not marked as default: override-type"
+							h.AssertStringContains(t, stdout.String(), expected)
+
+							h.AssertEq(t, metadata.BuildpackDefaultProcessType, "")
+						})
+					})
+
+					when("there is a web process", func() {
```

Comment:
We might want to add a test that when platform api < 0.6, we don't set default = true for web. But I'm not sure if it's really necessary.

Reply:
Please see https://github.com/buildpacks/lifecycle/pull/520/commits/7b24df17bb72f57cfba08d780795cbaba5aaddca.